Fix primary key behaviour in bulk_update

pfhayes opened the pull request:
Suppose you have a model class with a primary key.

```python
Base = declarative_base()
class User(Base):
  id = Column(BigInteger, primary_key=True)
  name = Column(String)
```

Previously, running
`bulk_update_mappings(User, {'id': 1, 'name': 'hello'})`
would emit the following:

```sql
UPDATE users SET id=1, name='hello' WHERE id=1
```

This is contrary to the stated behaviour, where primary keys are omitted
from the SET clause. Furthermore, this behaviour is harmful, as it
can cause the db engine to lock over-aggresively (at least in Postgres).

With this change, the emitted SQL is:

```sql
UPDATE users SET name='hello' WHERE id=1
```

zzzeek commented:
thanks very much for this!  I added an issue at https://bitbucket.org/zzzeek/sqlalchemy/issue/3451/bulk-update-adds-the-pk-to-the-set-clause as this was a pretty large oversight.  Ultimately, I modified the fix here to just not include these PK cols up where it gathers params in the first place, instead of adding them then removing them (see bcbfcca2360a58).  The "dict w/ generator" constructor saves on method calls vs. individual set operations.

pfhayes commented:
my pleasure! thanks for the speedy response

On Sat, Jun 13, 2015 at 8:48 PM, mike bayer <notifications@github.com>
wrote:

> thanks very much for this! I added an issue at
> https://bitbucket.org/zzzeek/sqlalchemy/issue/3451/bulk-update-adds-the-pk-to-the-set-clause
> as this was a pretty large oversight. Ultimately, I modified the fix here
> to just not include these PK cols up where it gathers params in the first
> place, instead of adding them then removing them (see bcbfcca
> <https://github.com/zzzeek/sqlalchemy/commit/bcbfcca2360a58ec1b5a9da1b2c45e8781904b82>).
> The "dict w/ generator" constructor saves on method calls vs. individual
> set operations.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zzzeek/sqlalchemy/pull/181#issuecomment-111766076>.
>
